src/values/acceleration-structure.cxx

```diff
@@ -0,0 +1,1484 @@
+            for (unsigned row = 0; row < numRows; ++row) {
+                result << Util::repeatedString(indent + 2, indentString) << "[ ";
+                for (unsigned col = 0; col < numCols - 1; ++col) {
+                    result << objectToWorld[row][col] << ", ";
+                }
+                result << objectToWorld[row][numCols - 1] << " ]" << std::endl;
+            }
+            result << Util::repeatedString(indent + 1, indentString) << "]" << std::endl;
+
+            result << Util::repeatedString(indent + 1, indentString) << "instance_mask = " << instanceMask << std::endl;
+
+            result << Util::repeatedString(indent + 1, indentString)
+                   << "points_to_acceleration_structure_id = " << accelerationStructure->id << std::endl;
+
+            return result.str();
+        }
+    };
+
+    class TriangleNode : public Node {
+    public:
+        const bool opaque;
+        const std::vector<glm::vec3> vertices;
+        const std::vector<unsigned> indices;
+
+        TriangleNode(const bool& opaque, const std::vector<glm::vec3>& vertices, const std::vector<unsigned>& indices)
+            : opaque(opaque),
+              vertices(vertices),
+              indices(indices) {};
+
+        NodeType type() {
+            return NodeType::Triangle;
+        }
+
+        std::string toString(const unsigned indent = 0, const std::string indentString = "") const {
+            std::stringstream result("");
+
+            result << Util::repeatedString(indent, indentString) << "triangle_node" << std::endl;
+
+            result << Util::repeatedString(indent + 1, indentString) << "opaque = " << (opaque ? "true" : "false")
+                   << std::endl;
+
+            result << Util::repeatedString(indent + 1, indentString) << "vertices = [" << std::endl;
+            for (unsigned i = 0; i < vertices.size() - 1; ++i) {
+                result << Util::repeatedString(indent + 2, indentString) << Util::glmVec3ToString(vertices[i]) << ","
+                       << std::endl;
+            }
+            result << Util::repeatedString(indent + 2, indentString)
+                   << Util::glmVec3ToString(vertices[vertices.size() - 1]) << std::endl;
+            result << Util::repeatedString(indent + 1, indentString) << "]" << std::endl;
+
+            result << Util::repeatedString(indent + 1, indentString) << "indices = [ ";
+            for (unsigned i = 0; i < indices.size() - 1; ++i) {
+                result << indices[i] << ", ";
+            }
+            result << indices[indices.size() - 1] << " ]" << std::endl;
+
+            return result.str();
+        }
+    };
+
+    class ProceduralNode : public Node {
+    public:
+        const bool opaque;
+        const glm::vec4 minBounds; // min x, y, z
+        const glm::vec4 maxBounds; // max x, y, z
+
+        ProceduralNode(const bool& opaque, const glm::vec4& minBounds, const glm::vec4& maxBounds)
```

Comment:
If you are passing a bool which won't be modified, I don't see the need to pass it by reference. You should often pass primitive types by value instead.